File: Minimum_Depth_of_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Recursion Version
#include <algorithm>
class Solution {
public:
    int minDepth(TreeNode* root) {
        return minDepthSearch(root, false);
    }
    int minDepthSearch(TreeNode* node, bool hasbrother){
        if(!node){
            if(!hasbrother) return 0;
            else return INT_MAX;
        }
        
        return 1 + min(minDepthSearch(node->left, node->right != nullptr), minDepthSearch(node->right, node->left != nullptr));
    }
};

File: Chapter_5_Tree/minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: TreeNode) -> int: 
        if not root:
            return 0
        depth = 1
        Q = deque([root])
        
        while len(Q) != 0:
            for _ in range(len(Q)):
                node = Q.popleft()
                if not node.left and not node.right:
                    return depth
                
                if node.left:
                    Q.append(node.left)
                if node.right:
                    Q.append(node.right)
                    
            depth += 1
                    
